addons/web/controllers/main.py

```diff
@@ -702,8 +702,9 @@ def backup(self, master_pwd, name, backup_format = 'zip'):
     @http.route('/web/database/restore', type='http', auth="none", methods=['POST'], csrf=False)
     def restore(self, master_pwd, backup_file, name, copy=False):
         try:
-            data = base64.b64encode(backup_file.read())
-            dispatch_rpc('db', 'restore', [master_pwd, name, data, str2bool(copy)])
+            data = ''
+            for chunk in iter(lambda: backup_file.read(8190), b''):
+                data += base64.b64encode(chunk)            dispatch_rpc('db', 'restore', [master_pwd, name, data, str2bool(copy)])
```

Comment:
base64 is a bytes -> bytes conversion, so `data` is more or less a bytestring (if dispatch_rpc needs a native string, the base64 content should be converted correctly or it's going to blow up badly in Python 3). 

Bytes are immutable, concatenating immutable string is [fundamentally quadratic](https://ericlippert.com/2013/06/24/string-concatenation-behind-the-scenes-part-two/). This code implicitly relies on a CPython optimisation to reclaim (amortised) linear behaviour. I would rather we avoid this issue, by using bytearray. The bytearray can be converted into regular bytes at the end of the loop.